#### Summary Functions & Grouping

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv('../data/NYC_Rolling_Sales_Dataset/nyc-rolling-sales.csv')

##### Convert Text to Numeric

In [3]:
df['SALE PRICE'] = pd.to_numeric(df['SALE PRICE'], errors='coerce')

##### Describe Numerical Columns

In [6]:
df['SALE PRICE'].describe()

count    6.998700e+04
mean     1.276456e+06
std      1.140526e+07
min      0.000000e+00
25%      2.250000e+05
50%      5.300000e+05
75%      9.500000e+05
max      2.210000e+09
Name: SALE PRICE, dtype: float64

##### Group wise Summary

In [8]:
# Convert BOTH columns to numbers properly
df['SALE PRICE'] = pd.to_numeric(df['SALE PRICE'], errors='coerce')
df['LAND SQUARE FEET'] = pd.to_numeric(df['LAND SQUARE FEET'], errors='coerce')
df['RESIDENTIAL UNITS'] = pd.to_numeric(df['RESIDENTIAL UNITS'], errors='coerce')

# Remove rows with any missing values
df_clean = df.dropna(subset=['SALE PRICE', 'LAND SQUARE FEET', 'RESIDENTIAL UNITS'])

# Now groupby will work!
result = df_clean.groupby('RESIDENTIAL UNITS')[['SALE PRICE', 'LAND SQUARE FEET']].mean().round(2)
print(result.head(10))

                   SALE PRICE  LAND SQUARE FEET
RESIDENTIAL UNITS                              
0                  2493212.64           7111.43
1                   656977.58           2364.07
2                   618955.55           2768.06
3                   666907.45           2441.95
4                   900805.63           2500.52
5                  1289258.59           2676.58
6                  1235253.85           2585.04
7                  3048610.05           2621.92
8                  2732408.65           2915.25
9                  3906328.27           3435.59


##### Categorize Sale Price

In [9]:
def price_category(price):
    if price > 1_000_000:
        return 'High'
    elif price > 500_000:
        return 'Medium'
    else:
        return 'Low'

df['PRICE_CATEGORY'] = df['SALE PRICE'].apply(price_category)
df[['SALE PRICE', 'PRICE_CATEGORY']].head()

SALE PRICE PRICE_CATEGORY
0   6625000.0           High
1         NaN            Low
2         NaN            Low
3   3936272.0           High
4   8000000.0           High

##### Combined Insight

In [10]:
df.groupby('PRICE_CATEGORY')['SALE PRICE'].count()

PRICE_CATEGORY
High      15592
Low       33555
Medium    20840
Name: SALE PRICE, dtype: int64

#### 📊 NYC Property Sales – Summary Insights (Module 3)

##### 📌 Dataset Overview
- **Total Properties:** 84,548 records  
- **Clean Data Used:** 69,987 properties (after removing missing values)  
- **Time Period:** NYC Rolling Sales data  

---

##### 1️⃣ Sale Price Statistics

**Summary Statistics (SALE PRICE):**

- **Count:** 69,987 properties  
- **Mean Price:** $1,276,456  
- **Median Price:** $530,000  
- **Standard Deviation:** $11,405,260  
- **Minimum:** $0  
- **Maximum:** $2,210,000,000  

##### 💡 Key Insights
- **Mean > Median:** A small number of extremely high-value properties significantly increase the average price.
- **Typical Property Value:** The median price of **$530,000** better represents the mid-market.
- **Wide Price Range:** Prices range from $0 (ownership transfers/gifts) to multi-billion-dollar commercial acquisitions.

---

##### 2️⃣ Property Types by Residential Units

| Residential Units | Avg Sale Price | Avg Land Area (sq ft) | Property Type |
|-------------------|--------------|-----------------------|---------------|
| 0 | $2.49M | 7,111 | Commercial |
| 1 | $657K | 2,364 | Single Family |
| 2 | $619K | 2,768 | Two-Family |
| 3 | $667K | 2,442 | Three-Family |
| 4 | $901K | 2,501 | Small Multi-Family |
| 5 | $1.29M | 2,677 | Apartment Building |
| 6 | $1.24M | 2,585 | Apartment Building |
| 7 | $3.05M | 2,622 | Large Apartment |
| 8 | $2.73M | 2,915 | Large Apartment |
| 9 | $3.91M | 3,436 | Large Apartment |

###### 💡 Key Insights
- **Commercial Properties (0 units):** Highest average prices and largest land footprints.
- **Common Residential Units (1–3):** Prices remain relatively stable between **$600K–$700K**.
- **Scaling Effect:** Property values rise sharply once the unit count exceeds **4 units**, indicating higher investment-grade assets.

---

##### 3️⃣ Price Category Distribution

##### 🏷️ Defined Price Categories
- **High:** > $1,000,000  
- **Medium:** $500,000 – $1,000,000  
- **Low:** < $500,000  

##### 📊 Distribution Breakdown
- 🏠 **Low Price:** 33,555 properties (**48%**)  
- 🏢 **Medium Price:** 20,840 properties (**30%**)  
- 💎 **High Price:** 15,592 properties (**22%**)  

##### 💡 Key Insights
- Nearly **half of the market** falls under the low-price category.
- High-value properties form a smaller but economically significant segment.
- The market is strongly skewed toward affordable and mid-range housing.

---